File: apps/androidtv/feature/settings/src/main/kotlin/dev/spidola/tv/feature/settings/SettingsSupport.kt
// SPDX-FileCopyrightText: 2026 Spidola contributors
// SPDX-License-Identifier: AGPL-3.0-or-later

@file:Suppress("MatchingDeclarationName") // Groups the settings slice's navigation + error bridge.

package dev.spidola.tv.feature.settings

import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import dev.spidola.tv.core.corekit.ActionableError
import dev.spidola.tv.core.corekit.ErrorAction
import dev.spidola.tv.core.designsystem.ActionableErrorView
import dev.spidola.tv.core.designsystem.SpidolaErrorButton

/**
 * The settings slice's navigation intents, wired by the app's composition root to the Navigation 3
 * back stack (TECH_SPEC §7). The slice depends on these lambdas, never on the app's route types, so
 * it stays free of sideways/upward dependencies (doctrine §3.1).
 */
data class SettingsNavigator(
    /** Opens the option list for one closed-set setting. */
    val openPicker: (SettingsPicker) -> Unit,
    val openDiagnostics: () -> Unit,
)

/**
 * Renders a corekit [ActionableError] through the designsystem [ActionableErrorView], wiring each
 * prescribed action to a handler (PRD §6.3). The bridge lives at the feature layer because it joins
 * the core error model (corekit) to the visual component (designsystem), and neither horizontal
 * layer should depend on the other — so each slice carries its own, rather than depending sideways
 * on another slice's copy.
 */
@Composable
internal fun ActionableErrorContent(
    error: ActionableError,
    onRetry: () -> Unit,
    onGoBack: () -> Unit,
    modifier: Modifier = Modifier,
) {
    fun button(action: ErrorAction): SpidolaErrorButton =
        SpidolaErrorButton(
            title = action.label,
            onClick =
                when (action) {
                    ErrorAction.RETRY -> onRetry
                    // Nothing on a settings screen is a form, so there is no input to send the
                    // viewer back to; going back is the honest offer.
                    ErrorAction.FIX_INPUT, ErrorAction.GO_BACK -> onGoBack
                },
        )
    ActionableErrorView(
        failureClass = error.failureClass,
        message = error.message,
        primary = button(error.primaryAction),
        others = error.otherActions.map(::button),
        modifier = modifier,
    )
}

File: apps/tvos/Packages/FeatureSettings/Sources/FeatureSettings/SettingsSupport.swift
// SPDX-FileCopyrightText: 2026 Spidola contributors
// SPDX-License-Identifier: AGPL-3.0-or-later

import CoreKit
import DesignSystem
import SwiftUI
import core_api

/// The settings slice's navigation intents, wired by the app's composition root to the state-driven
/// `NavigationStack` (TECH_SPEC §6). The slice depends on these closures, never on the app's typed
/// route enum, so it stays free of sideways/upward dependencies (doctrine §3.1).
@MainActor
public struct SettingsNavigator {
  /// Opens the picker for one closed-set setting. The root and the diagnostics screen share it —
  /// there is one picker screen in this slice, and the field it is about is its whole payload.
  public var openOptions: (_ field: SettingsField) -> Void
  public var openDiagnostics: () -> Void
  public var openAbout: () -> Void

  public init(
    openOptions: @escaping (SettingsField) -> Void,
    openDiagnostics: @escaping () -> Void,
    openAbout: @escaping () -> Void
  ) {
    self.openOptions = openOptions
    self.openDiagnostics = openDiagnostics
    self.openAbout = openAbout
  }
}

extension View {
  /// Renders a CoreKit `ActionableError` through the DesignSystem `ActionableErrorView`, wiring each
  /// prescribed action to a handler. The bridge lives at the feature layer because it joins the
  /// core error model (CoreKit) to the visual component (DesignSystem), and neither horizontal layer
  /// should depend on the other — which is also why each slice carries its own copy rather than
  /// importing a neighbour's.
  @MainActor
  func actionableError(
    _ error: ActionableError,
    retry: @escaping @MainActor () -> Void,
    goBack: @escaping @MainActor () -> Void
  ) -> some View {
    func button(_ action: ErrorAction) -> SpidolaErrorButton {
      switch action {
      case .retry: SpidolaErrorButton(title: action.label, action: retry)
      case .goBack: SpidolaErrorButton(title: action.label, action: goBack)
      // Nothing on a settings screen is typed in, so there is no input to go back and correct;
      // re-reading is the nearest honest action.
      case .fixInput: SpidolaErrorButton(title: ErrorAction.retry.label, action: retry)
      }
    }
    return ActionableErrorView(
      failureClass: error.failureClass,
      message: error.message,
      primary: button(error.primaryAction),
      others: error.otherActions.map(button))
  }
}
